### PR TITLE
[Relay][Legalize][ARM_CPU] Handling NHWC layout for arm_cpu.

### PR DESCRIPTION
**Pull request by anijain2305**
Relevant Issue - #2519

Currently, the TFLite parsed models fail for ARM devices. The reason is TFLite models are NHWC and we don't support NHWC layout in ARM. However, TF models (which are also NHWC), are able to run on ARM devices. This is because of TF parser adds layout transforms while parsing the graph (which is not a good idea as parsers should retain the framework layout).

As a first step, this PR adds a legalize function for conv2d for arm_cpu to add transposes before and after to use the NCHW layout. This gets the TFLite models working for ARM cpus.

A follow-up work requires TF parser cleanup. Parsers should keep the frontend layout. So, we can add legalize for all the targets to fall back to NCHW for conv and then cleanup TF parser. (Let me know if I should open up a new Issue for tracking the tasks).

@yzhliu @FrozenGene @apivovarov @kevinthesun @yongwww

### CODE REVIEW

**Comment by FrozenGene** [comment]
Inserting `transpose` is not good solution IMO. As demonstrated, which has big impact performance when conv2d is executed fast. Besides us, the community also report this thing too. https://discuss.tvm.ai/t/quantization-autotvm-performance-degradation-due-to-transpose-ops-when-nchw-layout-is-used/3561. I think we should support NHWC layout of spatial pack schedule in ARM CPU instead of inserting transpose op. Morever, @jackwish has investigated NHWC / NCHW layout on ARM CPU advantages / disadvantages internally. If we don't take NHWC layout, we can not get the best performance and beat QNNPACK. 

In summary, we should support NHWC layout like here: https://github.com/dmlc/tvm/blob/master/topi/python/topi/arm_cpu/bitserial_conv2d.py#L264 and it is not difficult and only need our minor work. Unified one NCHW sounds good but is not best in practical if we want to get the best performance (which is one key factor to let users use TVM or not).

**Comment by anijain2305** [comment]
@FrozenGene I agree writing schedule for NHWC is a good long-term solution. However, for the time being, can this be merged in? Current state is bad because we fail compilation.

**Comment by FrozenGene** [comment]
I am busy in our internal DSP product. I can not make sure whether @jackwish could have interest and have time to help it, I think we could contribute it back.

**Comment by anijain2305** [comment]
Ok, in that case,  I would suggest current PR as an intermediate step.

Performance is definitely one of the key factors for using TVM. But, equally important is portability, both across frameworks and HW devices. It is fine to be in a temporary stage where things are functional and not performant as compared to a stage where we are neither functional nor performant. Currently, we fall into latter bucket for TFLite models, which makes it harder to convince edge device vendors (who like TFLite) to try TVM.

File: topi/python/topi/nn/conv2d.py [hunk 1]
```diff
@@ -71,6 +71,28 @@ def conv2d(input, filter, strides, padding, dilation, layout='NCHW', out_dtype=N
...
 
 
+@tvm.target.generic_func
+def conv2d_legalize(attrs, inputs, arg_dtypes, F):
+    """Legalizes Conv2D op.
+    Parameters
+    ----------
+    attrs : nnvm.top.AttrDict or tvm.attrs.Attrs
+        Attributes of current convolution
+    inputs : nnvm.symbol or tvm.relay.Expr
```
Comment (wweic) [inline topi/python/topi/nn/conv2d.py]: Arn't we going to discontinue nnvm support?

Comment (anijain2305) [reply topi/python/topi/nn/conv2d.py]: Thanks! Yes, I have made it work only for Relay. Changed the comments to reflect that.

**Comment by jackwish** [comment]
Yes, I may have time to look into this, if the *internal license* won't be a problem...

**Comment by jackwish** [comment]
> I am busy in our internal DSP product. I can not make sure whether @jackwish could have interest and have time to help it, I think we could contribute it back.

We may need some internal discussions on this...

File: topi/python/topi/arm_cpu/conv2d.py [hunk 2]
```diff
@@ -783,3 +784,31 @@ def _alter_conv2d_layout_arm(attrs, inputs, tinfos, F):
             # currently we only have contrib_spatial_pack and direct template
             # add more schedule templates.
             return None
+
+@conv2d_legalize.register("arm_cpu")
+def _conv2d_legalize(attrs, inputs, arg_types, F):
+    if F.__name__ != 'tvm.relay.op':
+        return None
+    if attrs['data_layout'] == 'NHWC':
```
Comment (jackwish) [inline topi/python/topi/arm_cpu/conv2d.py]: Would it be better to warn user about this?

**Comment by anijain2305** [comment]
> Thanks for ping. Are we using this to make it possible to test models (I mean something like mobilenet)?

Yes

File: topi/python/topi/arm_cpu/conv2d.py (hunk 2 above)
Comment (anijain2305) [reply topi/python/topi/arm_cpu/conv2d.py]: I can add a warning here.

**Comment by kevinthesun** [comment]
@FrozenGene I think this PR doesn't introduce transpose by replacing conv2d NHWC directly conv2d NCHW. This partially resolve our issue. In the long term, we might need https://github.com/dmlc/tvm/issues/3670 to better manage graph level layout conversion. For topi schedule, I suggest we only add schedules for new layout when obvious performance benefit is shown. Otherwise we can just leverage existing layout schedules and convert node into that layout.

**Comment by FrozenGene** [comment]
> @FrozenGene I think this PR doesn't introduce transpose by replacing conv2d NHWC directly with conv2d NCHW. This partially resolves our issue. In the long term, we might need #3670 to better manage graph level layout conversion. For topi schedule, I suggest we only add schedules for new layout when obvious performance benefit is shown. Otherwise we can just leverage existing layout schedules and convert nodes into that layout.

Yeah, we have investigated NHWC and NCHW layout in quantization model deeply and find NHWC could make us get better performance, otherwise we will introduce `pack`  operation to get the same effect (better locality)  of NHWC. However, when we execute the quantized model, we care the performance very much. The `pack` cost time we can not  ignore. @jackwish maybe could share his investigation result more.

**Comment by kevinthesun** [comment]
@FrozenGene I see. I think it's no harm that we improve layout support and topi schedule at the same time. This PR can serve as a short term solution for arm_cpu fp32.

File: topi/python/topi/arm_cpu/conv2d.py (hunk 2 above)
Comment (anijain2305) [reply topi/python/topi/arm_cpu/conv2d.py]: Added at L803

**Comment by kevinthesun** [comment]
Thanks!

File: topi/python/topi/nn/conv2d.py [hunk 3]
```diff
@@ -71,6 +71,28 @@ def conv2d(input, filter, strides, padding, dilation, layout='NCHW', out_dtype=N
...
+    inputs : list of tvm.relay.Expr
+        The args of the Relay expr to be legalized.
+    arg_dtypes : list of types
+        List of types of input arguments
+    F: symbol
+        The context, can be either nnvm.sym or relay.op
+    Note
+    ----
+    Unlike other TOPI functions, this function operates on both graph level and operator level,
+    so we have to pass 'F' to make it support our two versions of graph IR, NNVM and Relay.
```
Comment (vinx13) [inline topi/python/topi/nn/conv2d.py]: @anijain2305 why are NNVM and `F` argument still mentioned here?

Comment (anijain2305) [reply topi/python/topi/nn/conv2d.py]: Ah, thanks for pointing out. I copied the description from other functions which have this comment. Will send a separate PR to clean up the comments.

**Comment by jackwish** [comment]
Drafted the ARM NHWC schedule PR https://github.com/dmlc/tvm/pull/3859, if you guys have interest.
